# Sklearn dataset loading utilities

scikit-learn Python library comes with a few small standard datasets that do not require to download any file from some external website, as they are available in sci-kit learn installation by executing `sklearn.datasets` package.

### General dataset API

There are three main kinds of dataset interfaces that can be used to get datasets depending on the desired type of dataset.

![General Dataset API](images/general_dataset_api.png)

 * **The dataset loaders** 
 
 Can be used to load small standard datasets, they are also called _the Toy datasets_, and I will talk about more in the next sections of this document.

 *  **The dataset fetchers**
 
 They can be used to download and load larger datasets. For further information refer to the _Real world datasets_ section in [scikit learn site](https://scikit-learn.org/stable/datasets/real_world.html).


 * **The dataset generation functions** 
 
 They can be used to generate controlled synthetic datasets. Find more information in the _Generated datasets_ section in [scikit learn site](https://scikit-learn.org/stable/datasets/sample_generators.html).



 The datasets also contain a full description in their `DESCR` attribute and some contain `feature_names` and `target_names`. See the dataset descriptions below for details.

#### Toy datasets

scikit-learn comes with a few small standard datasets that do not require to download any file from some external website.

They can be loaded using the following functions:

Dataset | Description
--- | ---
**load_boston**(*[, return_X_y]) | `Load and return the boston house-prices dataset (regression).`
**load_iris**(*[, return_X_y, as_frame]) | `Load and return the boston house-prices dataset (regression).`
**load_diabetes**(*[, return_X_y, as_frame]) | `Load and return the diabetes dataset (regression).`
**load_digits**(*[, n_class, return_X_y, as_frame]) | `Load and return the digits dataset (classification).`
**load_linnerud**(*[, return_X_y, as_frame]) | `Load and return the physical excercise linnerud dataset.`
**load_wine**(*[, return_X_y, as_frame]) | `Load and return the wine dataset (classification).`
**load_breast_cancer**(*[, return_X_y, as_frame]) | `Load and return the breast cancer wisconsin dataset (classification).`

Have in mind that these datasets are useful to quickly illustrate the behavior of the various algorithms implemented in scikit-learn, eventhough, they are considered too small to be representative of real world machine learning tasks.

In [1]:
import pandas as pd
from sklearn.datasets import load_wine

Now that we have loaded a toy dataset from sklearn API by applying the function `load_wine()`, we store it inside the variable _data_

In [2]:
wine = load_wine()

Next let's make use of _shape_ in order to inpect how many colums and rows it has.

In [3]:
print(f'The wine dataset contains {wine.data.shape[0]} rows and {wine.data.shape[1]} columns')

The wine dataset contains 178 rows and 13 columns


Looking at the dataset's data type, we notice that _sklearn.utils.Bunch_ is returned, For more information about this Bunch object go to this [link](https://scikit-learn.org/stable/modules/generated/sklearn.utils.Bunch.html#sklearn.utils.Bunch)

In [4]:
print(type(wine))

<class 'sklearn.utils.Bunch'>


So far we know that our wine toy dataset is comprised by 178 rows and 13 columns, but we haven't still made a first sight to it, so for that let's use the _DESCR_ attribute as talked before, this works in a similar way as pandas _describe()_ function, but it provides more detailed information regarding the dataset.

In [5]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

By applying the _key()_ function, we'll have access to a list of attributes available to apply on this dataset.

In [6]:
wine.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

The first one is `data`, which returns an array of the data itself contained in the dataset.

In [7]:
wine.data

array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
        1.185e+03],
       ...,
       [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
        8.350e+02],
       [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
        8.400e+02],
       [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
        5.600e+02]])

The fourth one gives us the name of the classes of wine, means the target features.

In [8]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

And the last one the name of all features in the dataset

In [13]:
wine.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

### Converting Sklearn dataset in pandas dataframe

Though the loaded dataset from sklearn API is ready to be used for machine learning algorithms, it is also useful to convert it into pasdas dataframe for other purposes.

In [9]:
df_wine = pd.DataFrame(data=wine.data,columns=wine.feature_names)
df_wine.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  
0                          3.92   1065.0  
1                          3.40   1050.0  
2                          3.17   1185.0  
3                          3.45   1480.0  
4                          2.93    735.0

In [14]:
df_wine.describe()

alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       color_intensity         hue  od280/od315_of_diluted_wines      proline  
count       178.000000  178.000000                    178.000000   178.000000  
mean          5.058090    0.957449                      2.611685   746.893258  
std           2.318286    0.228572                      0.709990   314.907474  
min           1.280000    0.480000                      1.270000   278.000000  
25%           3.220000    0.782500                      1.937500   500.500000  
50%           4.690000    0.965000                      2.780000   673.500000  
75%           6.200000    1.120000                      3.170000   985.000000  
max          13.000000    1.710000                      4.000000  1680.000000

One way or another will be the starting point for our EDA or data preprosessing for Machine Learning. In this tutorial we have learned how to use pre-loaded datasets available in sklearn API, if you are interested on cloning this notebook find it in my [GitHub repo](https://github.com/fvgm-spec/ML/blob/main/notebooks/Sklearn%20toy%20datasets.ipynb) or if you need more documentation find it in [sklearn site](https://scikit-learn.org/stable/datasets.html)

### To be continued...